api/oneclick.py

```diff
@@ -0,0 +1,453 @@
+)
+
+from core.util.http import (
+    BadResponseException,
+)
+
+
+class OneClickAPI(BaseOneClickAPI, BaseCirculationAPI):
+
+    NAME = "OneClick"
+    
+    EXPIRATION_DATE_FORMAT = '%m-%d-%y'
+    EXPIRATION_DEFAULT = datetime.timedelta(days=21)
+
+    # How long we should go before syncing our internal Patron record
+    # with Millenium.
+    #MAX_STALE_TIME = datetime.timedelta(hours=12)
+
+    log = logging.getLogger("OneClick Patron API")
+
+
+    def __init__(self, *args, **kwargs):
+        super(OneClickAPI, self).__init__(*args, **kwargs)
+        self.bibliographic_coverage_provider = (
+            OneClickBibliographicCoverageProvider(
+                self._db, oneclick_api=self
+            )
+        )
+
+
+    def checkin(self, patron, pin, licensepool):
+        """
+        Allow a patron to return an ebook or audio before its due date.
+
+        :param patron: a Patron object for the patron who wants to return the book.
+        :param pin: The patron's password (not used).
+        :param licensepool: The Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :return True on success, raises circulation exceptions on failure.
+        """
+        (patron_oneclick_id, item_oneclick_id) = self.validate_input(patron, licensepool)
+
+        resp_dict = self.circulate_item(patron_id=patron_oneclick_id, item_id=item_oneclick_id, return_item=True)
+
+        if resp_dict == {}:
+            self.log.debug("Patron %s/%s returned item %s.", patron.authorization_identifier, 
+                patron_oneclick_id, item_oneclick_id)
+            return True
+
+        # should never happen
+        raise CirculationException("Unknown error %s/%s checking in %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id)
+
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """
+        Associate an ebook or audio with a patron.
+
+        :param patron: a Patron object for the patron who wants to check out the book.
+        :param pin: The patron's password (not used).
+        :param licensepool: The Identifier of the book to be checked out is 
+        attached to this licensepool.
+        :param internal_format: Represents the patron's desired book format.  Ignored for now.
+
+        :return LoanInfo on success, None on failure
+        """
+        (patron_oneclick_id, item_oneclick_id) = self.validate_input(patron, licensepool)
+
+        resp_dict = self.circulate_item(patron_id=patron_oneclick_id, item_id=item_oneclick_id, return_item=False)
+
+        if not resp_dict or ('error_code' in resp_dict):
+            return None
+
+        self.log.debug("Patron %s/%s checked out item %s with transaction id %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id, resp_dict['transactionId'])
+
+        today = datetime.datetime.now()
+        expires = today + self.EXPIRATION_DEFAULT
```

Comment:
How certain are we that loans are always 21 days long? Is this a policy across One Click?